core/finality-grandpa/primitives/src/lib.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives for GRANDPA integration, suitable for WASM compilation.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate substrate_primitives;
+extern crate sr_primitives;
+extern crate parity_codec;
+
+#[macro_use]
+extern crate parity_codec_derive;
+
+#[macro_use]
+extern crate sr_api;
+
+use substrate_primitives::AuthorityId;
+use sr_primitives::traits::{Block as BlockT, DigestFor, NumberFor};
+
+/// A scheduled change of authority set.
+#[cfg_attr(feature = "std", derive(Debug, PartialEq))]
+#[derive(Clone, Encode, Decode)]
+pub struct ScheduledChange<N> {
+	/// The new authorities after the change, along with their respective weights.
+	pub next_authorities: Vec<(AuthorityId, u64)>,
+	/// The number of blocks to delay.
+	pub delay: N,
+}
+
+/// WASM function call to check for pending changes.
+pub const PENDING_CHANGE_CALL: &str = "grandpa_pending_change";
+/// WASM function call to get current GRANDPA authorities.
+pub const AUTHORITIES_CALL: &str = "grandpa_authorities";
+
+decl_apis! {
+	/// APIs for integrating the GRANDPA finality gadget into runtimes.
+	///
+	/// This is primarily used for negotiating authority-set changes for the
+	/// gadget. GRANDPA uses a signalling model of changing authority sets:
+	/// changes should be signalled with a delay of N blocks, and then automatically
+	/// applied in the runtime after those N blocks have passed.
+	///
+	/// The consensus protocol will coordinate the handoff externally.
+	pub trait Api<B: BlockT> {
```

Comment:
There's a convention to use a `CAPS_CASE` version of the `ApiTraitName` as an `ApiId` ("API identifier") that allows for extensible runtime versioning for each API:

```
pub mod id {
	use super::ApiId;
	pub const BLOCK_BUILDER: ApiId = *b"blkbuild";
	pub const TAGGED_TRANSACTION_QUEUE: ApiId = *b"validatx";
	pub const METADATA: ApiId = *b"metadata";
}

decl_apis! {
	pub trait Core<Block: BlockT, AuthorityId> { ... }
	pub trait Metadata<Data> { ... }
	pub trait TaggedTransactionQueue<Block: BlockT> { ... }
	pub trait BlockBuilder<Block: BlockT> ExtraClientSide <OverlayedChanges> { ... }
}
```

 A terse name like `Api` makes this impractical or error prone. I would suggest `Grandpa`/`GrandpaInterface` or some other such name to prevent clashes and confusion.

Reply:
Some documentation on the `decl_apis!` macro would be helpful; there is no way to know about this convention otherwise.

On the naming, I intended the name to be used like this:

```rust
extern crate grandpa_primitives as grandpa;

fn uses_api<T: grandpa::Api>(T) { }
```